File: ios/Gildongmu/AppConfig.swift
import Foundation

/// 앱 전역 설정. base URL은 릴리스 고정, 디버그는 스킴 환경변수로 로컬 dev 전환(spec §6).
enum AppConfig {
    static var apiBaseURL: URL {
        #if DEBUG
        if let override = ProcessInfo.processInfo.environment["GILDONGMU_API_BASE_URL"],
           let url = URL(string: override) {
            return url
        }
        #endif
        return URL(string: "https://gildongmu.vercel.app")!
    }

    /// nmap 딥링크 필수 appname(웹 NEXT_PUBLIC_APP_IDENTIFIER와 동일값)
    static let appIdentifier = "space.dodoplanet.gildongmu"

    /// 아직 정식 출시하지 않은 안내 수단(자동차·대중교통·간략 단독 진입점)의 봉인.
    /// 도보 상세 안내는 정식 출시(2026-08-15, spec 2026-08-15-walk-guidance-ship-design.md)로
    /// 졸업해 이 플래그를 보지 않는다 — 이탈 판정 방위 축도 켠 채로 나간다(위원장 판정:
    /// 실험판과 정식판의 안내 동작을 같게 유지해 앞으로의 실보행 판정이 정식판에 그대로
    /// 적용되게 한다. 축을 끄면 갈림길 오진 원증상(A6)이 남는데 그쪽이 헛경고보다 위험하다).
    ///
    /// **값을 손으로 고치지 않는다. 빌드 구성이 정한다**(2026-08-04 전환). 봉인의 판정
    /// 축은 플래그 참조 목록이 아니라 **세션을 시작시키는 호출 전수**다(spec §3.2 표) —
    /// 신규 진입점을 만들면 그 표와 `guidance-gate-drift.test.ts`를 함께 갱신하는 것이
    /// 계약이다. 남은 셋을 하나로 묶는 이유: 셋의 운명이 같다(실주행·실승차 판정 대기).
    /// 자동차가 먼저 졸업하는 상황이 실제로 오면 그때 쪼갠다.
    ///
    /// ⚠ 수단이 실주행·실승차 판정을 통과해 정식 출시할 때는 여기서 그 수단의 검사를
    /// 삭제한다(플래그 졸업). 플래그가 쌓이지 않게 하는 것이 이 방식의 유일한 관리 포인트다.
    #if EXPERIMENTAL
    static let experimentalGuidanceEnabled = true
    #else
    static let experimentalGuidanceEnabled = false
    #endif

    /// 웹 개인정보 처리방침 URL(현재 앱 언어 로케일). 동의 화면·설정이 공유한다.
    static var privacyPolicyURL: URL {
        apiBaseURL.appending(path: "\(AppLanguage.current)/privacy")
    }
}

File: android/app/src/main/kotlin/space/dodoplanet/gildongmu/AppConfig.kt
package space.dodoplanet.gildongmu

import android.content.Context
import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import space.dodoplanet.gildongmu.kit.APIClient
import space.dodoplanet.gildongmu.kit.SearchService
import space.dodoplanet.gildongmu.location.CurrentAddressStore
import space.dodoplanet.gildongmu.location.AndroidLocationSource
import space.dodoplanet.gildongmu.location.AndroidPermissionGate
import space.dodoplanet.gildongmu.location.LocationStore
import space.dodoplanet.gildongmu.location.ManualLocationStore
import space.dodoplanet.gildongmu.storage.SharedPreferencesStore
import space.dodoplanet.gildongmu.net.HttpUrlConnectionTransport

/**
 * 앱 전역 설정(iOS `AppConfig` 미러). 실험 기능은 플래그 값을 손으로 고치는 것이 아니라
 * 빌드 구성(`experimental`)으로 가른다. 검증되면 참조 자체를 지운다(항상 참 상수를 남기지 않는다).
 */
object AppConfig {
    val experimentalGuidanceEnabled: Boolean = BuildConfig.EXPERIMENTAL

    /** 탭 바 순서(iOS `experimentalTabOrderEnabled` 미러, K1 ① 위원장 판정 2026-08-23 실험판 판정 대기). */
    val experimentalTabOrderEnabled: Boolean = BuildConfig.EXPERIMENTAL

    /** 서버 base URL(대외 정본 도메인). 서버 계약 변경 0 — 기존 라우트만 부른다. */
    const val API_BASE_URL = "https://gildongmu.dodoplanet.space"

    /** nmap 딥링크 필수 appname(웹 NEXT_PUBLIC_APP_IDENTIFIER·iOS와 동일값). */
    const val APP_IDENTIFIER = "space.dodoplanet.gildongmu"

    /** :kit 판정 계층 + :app 전송 구현의 결합점. 화면들이 공유한다. */
    val apiClient: APIClient by lazy { APIClient(API_BASE_URL, HttpUrlConnectionTransport()) }

    private lateinit var app: Context

    /** `GildongmuApplication.onCreate`가 1회 부른다. */
    fun attach(context: Context) {
        app = context.applicationContext
    }

    /** 권한 대기 슬롯 소유자(spec §4). `MainActivity`가 손 둘을 등록·해제한다. */
    val permissionGate: AndroidPermissionGate by lazy { AndroidPermissionGate(app) }

    /** 현재 위치 공유 스토어 — 화면마다 `LocationManager`를 만들지 않는다. */
    val locationStore: LocationStore by lazy { LocationStore(AndroidLocationSource(app), permissionGate, log = { Log.i("Location", it) }) }

    /** 현재 위치 주소 캐시(표시줄, spec §12-4) — 좌표당 1회 역지오코딩. */
    val currentAddressStore: CurrentAddressStore by lazy { CurrentAddressStore(locationStore, SearchService(apiClient)) }

    /** 앱 수명 코루틴 스코프(hydration 등 프로세스 수명 작업). */
    val appScope: CoroutineScope by lazy { CoroutineScope(SupervisorJob() + Dispatchers.Default) }

    /** 수동 위치 런타임 정본(spec §13-1) — `GildongmuApplication`이 IO에서 `hydrate()`를 띄운다. */
    val manualLocationStore: ManualLocationStore by lazy { ManualLocationStore(SharedPreferencesStore(app)) }
}
